src/Features/Core/Portable/CodeAnalysisSuggestions/CodeAnalysisSuggestionsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,183 @@
+                    totalViolationsForCategory += totalInCurrentProject;
+                }
+            }
+
+            if (nestedActionsBuilder.Count == 0)
+                continue;
+
+            Debug.Assert(totalViolationsForCategory > 0);
+            totalViolations += totalViolationsForCategory;
+
+            // Add code action to Configure severity for the entire 'Category'
+            var categoryConfigurationAction = ConfigureSeverityLevelCodeFixProvider.CreateBulkSeverityConfigurationCodeAction(category, document.Project);
+            nestedActionsBuilder.Add(categoryConfigurationAction);
+
+            // {0} ({1})
+            var categoryBasedTitle = string.Format(FeaturesResources.Code_analysis_improvements_category_based_title,
+                category, totalViolationsForCategory);
+            var action = CodeAction.Create(categoryBasedTitle, nestedActionsBuilder.ToImmutableAndClear(), isInlinable: false);
+            actionsBuilder.Add(action);
+        }
+
+        if (actionsBuilder.Count > 0)
+        {
+            var disablingAction = CodeAction.Create(FeaturesResources.Do_not_show_Code_analysis_improvements,
+                createChangedSolution: cancellationToken =>
+                {
+                    var globalOptions = document.Project.Solution.Services.ExportProvider.GetExports<IGlobalOptionService>().Single().Value;
+                    globalOptions.SetGlobalOption(CodeAnalysisSuggestionsOptionsStorage.ShowCodeAnalysisSuggestionsInLightbulb, false);
+                    return Task.FromResult(document.Project.Solution);
+                },
+                equivalenceKey: nameof(FeaturesResources.Do_not_show_Code_analysis_improvements));
+            actionsBuilder.Add(disablingAction);
+        }
+
+        return (actionsBuilder.ToImmutable(), totalViolations);
+
+        static (DiagnosticData, Document?) GetPreferredDiagnosticAndDocument(ImmutableArray<DiagnosticData> diagnostics, Document document)
+        {
+            (DiagnosticData diagnostic, DocumentId? documentId)? preferredDiagnosticAndDocumentId = null;
+            foreach (var diagnostic in diagnostics)
+            {
+                if (diagnostic.DocumentId == document.Id)
+                {
+                    return (diagnostic, document);
+                }
+                else if (!preferredDiagnosticAndDocumentId.HasValue &&
+                    diagnostic.DocumentId?.ProjectId == document.Project.Id)
+                {
+                    preferredDiagnosticAndDocumentId = (diagnostic, diagnostic.DocumentId);
+                }
+            }
+
+            if (preferredDiagnosticAndDocumentId.HasValue)
+            {
+                return (preferredDiagnosticAndDocumentId.Value.diagnostic,
+                    document.Project.GetDocument(preferredDiagnosticAndDocumentId.Value.documentId!));
+            }
+
+            return (diagnostics.First(), null);
+        }
```

Comment:
this feels like it could be done with two loops.  first loop checks for exact docid match and returns first it finds.  second loop looks ahain and loops for projectId match and returns first it finds.  much simpler.  needing to do a single pass doesn't seem valuable.

Reply:
Done